core/scale/big_fixed_integers.hpp

```diff
@@ -0,0 +1,180 @@
+
+  template <typename Stream,
+            typename N,
+            typename = std::enable_if_t<Stream::is_encoder_stream>,
+            typename = std::enable_if_t<std::numeric_limits<N>::is_integer>>
+  Stream &operator<<(Stream &stream, Compact<N> compact) {
+    scale::CompactInteger n = compact;
+    stream << n;
+    return stream;
+  }
+
+  template <typename Stream,
+            typename N,
+            typename = std::enable_if_t<Stream::is_decoder_stream>,
+            typename = std::enable_if_t<std::numeric_limits<N>::is_integer>>
+  Stream &operator>>(Stream &stream, Compact<N> &compact) {
+    scale::CompactInteger n;
+    stream >> n;
+    compact.number = n.convert_to<N>();
+    return stream;
+  }
+
+  // intended to wrap unwrapped integer arguments to encoding functions to
+  // indicated they should be encoded with compact encoding
+  template <typename T>
+  struct EncodeAsCompact {
+    explicit EncodeAsCompact(T &number) : number{number} {}
+    T &number;
+  };
+
+  // marker for big integer types
+  template <typename T>
+  struct BigIntegerTraits {
+    static constexpr bool value = false;
+
+    static constexpr size_t BIT_SIZE = sizeof(T) * 8;
+  };
+
+  using uint128_t = boost::multiprecision::uint128_t;
+  using uint256_t = boost::multiprecision::uint256_t;
+
+  template <>
+  struct BigIntegerTraits<uint128_t> {
+    static constexpr size_t BIT_SIZE = 128;
+    static constexpr bool value = true;
+  };
+
+  template <>
+  struct BigIntegerTraits<uint256_t> {
+    static constexpr size_t BIT_SIZE = 256;
+    static constexpr bool value = true;
+  };
+
+  // big integers should only be encoded wrapped in Fixed, EncodeAsFixed,
+  // Compact, or EncodeAsCompact
+  template <typename Stream,
+            typename N,
+            typename = std::enable_if_t<Stream::is_decoder_stream>,
+            typename = std::enable_if_t<BigIntegerTraits<N>::value>>
+  Stream &operator>>(Stream &, N &) {
+    static_assert(!BigIntegerTraits<N>::value,  // just for dependent context
+                  "You need to wrap your number in EncodeAsFixed or "
+                  "EncodeAsCompact to clarify the encoding");
+  }
+
+  template <typename Stream,
+            typename N,
+            typename = std::enable_if_t<Stream::is_decoder_stream>,
+            typename = std::enable_if_t<BigIntegerTraits<N>::value>>
+  Stream &operator>>(Stream &stream, EncodeAsFixed<N> fixed) {
+    for (size_t i = 0; i < BigIntegerTraits<N>::BIT_SIZE; i += 8) {
+      fixed.number |= N(stream.nextByte()) << i;
+    }
+    return stream;
+  }
+
+  template <typename Stream,
+            typename N,
+            typename = std::enable_if_t<Stream::is_encoder_stream>,
+            typename = std::enable_if_t<BigIntegerTraits<N>::value>>
+  Stream &operator<<(Stream &stream, EncodeAsFixed<N> fixed) {
+    constexpr size_t bits = BigIntegerTraits<N>::BIT_SIZE;
+    for (N i = 0; i < bits; i += 8) {
+      stream << fixed & 0xFF;
+      fixed >>= 8;
+    }
```

Comment:
There are [`export_bits` and `import_bits`](https://www.boost.org/doc/libs/1_65_0/libs/multiprecision/doc/html/boost_multiprecision/tut/import_export.html) functions.  
[Example of custom output iterator](https://github.com/filecoin-project/cpp-filecoin/blob/master/core/vm/fvm/ubig128.hpp#L29-L47).

Maybe remove `BIT_SIZE` if it's not used elsewhere.

Reply:
The documentation was rather unclear about the exact format the bits are imported in (but in practice it's the correct lsb order) so I avoided it. However now these functions are used for all kinds of integers, not just from boost multiprecision.

Reply:
They accept `chunk_bit_width` and `bool lsb_or_msb`.

`BigIntegerTraits` is only implemented for multiprecision types.  
Standard integers are already encoded as fixed.